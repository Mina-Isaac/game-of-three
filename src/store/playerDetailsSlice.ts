import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { RootState } from "./store";

export interface Player {
  type: "human" | "cpu";
  name: string;
}
export interface PlayerState {
  player1: undefined | Player;
  player2: undefined | Player;
}

export const initialState: PlayerState = {
  player1: undefined,
  player2: undefined,
};

export const playerSlice = createSlice({
  name: "players",
  initialState,
  reducers: {
    initializePlayer: (
      state,
      action: PayloadAction<Player & { playerKey: keyof PlayerState }>
    ) => {
      const { playerKey, ...details } = action.payload;
      state[playerKey] = { ...details };
    },
  },
});

export const { initializePlayer } = playerSlice.actions;

// Selectors
export const selectPlayer1 = (state: RootState) => state.player.player1;
export const selectPlayer2 = (state: RootState) => state.player.player2;

export default playerSlice.reducer;
