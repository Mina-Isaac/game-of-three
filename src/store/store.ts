import { configureStore, combineReducers } from "@reduxjs/toolkit";
import playerReducer from "./playerDetailsSlice";
import gameReducer from "./gameSlice";
import {
  createStateSyncMiddleware,
  initStateWithPrevTab,
  withReduxStateSync,
} from "redux-state-sync";

const stateSyncMiddleware = createStateSyncMiddleware();

export const rootReducer = combineReducers({
  player: playerReducer,
  game: gameReducer,
});

const reducer = withReduxStateSync(rootReducer);

export const store = configureStore({
  reducer,
  middleware: [stateSyncMiddleware],
  devTools: process.env.NODE_ENV === 'development'
});
initStateWithPrevTab(store);

export type RootState = ReturnType<typeof rootReducer>;
