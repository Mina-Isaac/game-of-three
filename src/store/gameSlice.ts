import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { PlayerState } from "./playerDetailsSlice";
import { RootState } from "./store";

interface GameState {
  startingNumber: number | undefined;
  baseNumber: number | undefined;
  validMove: -1 | 0 | 1 | undefined;
  vs: "human" | "cpu" | undefined;
  lastPlayer: keyof PlayerState;
  canPlay: { player1: boolean; player2: boolean };
  winner: keyof PlayerState | undefined;
}

export const initialState: GameState = {
  startingNumber: undefined,
  baseNumber: undefined,
  validMove: undefined,
  vs: undefined,
  lastPlayer: "player1",
  canPlay: { player1: false, player2: true },
  winner: undefined,
};

export const gameSlice = createSlice({
  name: "game",
  initialState,
  reducers: {
    initializeGame: (state, action: PayloadAction<number>) => {
      const { payload } = action;
      state.startingNumber = payload;
      state.baseNumber = payload;
      state.lastPlayer = "player1";
      state.canPlay = { player1: false, player2: true };
      if (payload % 3 === 0) state.validMove = 0;
      else if (payload % 3 === 1) state.validMove = -1;
      else state.validMove = 1;
    },
    playNextMove: (state) => {
      const newBase = (state.baseNumber! + state.validMove!) / 3;
      state.baseNumber = newBase;
      if (newBase % 3 === 0) state.validMove = 0;
      else if (newBase % 3 === 1) state.validMove = -1;
      else state.validMove = 1;
      const { player1, player2 } = state.canPlay;
      state.canPlay = { player1: !player1, player2: !player2 };
      state.lastPlayer =
        (state.lastPlayer === "player1" && "player2") || "player1";
      if (newBase === 1) state.winner = state.lastPlayer;
    },
    setVs: (state, action: PayloadAction<"human" | "cpu" | undefined>) => {
      state.vs = action.payload;
    },
    startNewGame: (state) => {
      state.baseNumber = undefined;
      state.validMove = undefined;
      state.winner = undefined;
    },
  },
});

export const {
  initializeGame,
  playNextMove,
  setVs,
  startNewGame,
} = gameSlice.actions;

// Selectors
export const selectStartingNumber = (state: RootState) =>
  state.game.startingNumber;
export const selectBaseNumber = (state: RootState) => state.game.baseNumber;
export const selectVs = (state: RootState) => state.game.vs;
export const selectValidMove = (state: RootState) => state.game.validMove;
export const selectLastPlayer = (state: RootState) => state.game.lastPlayer;
export const SelectWinner = (state: RootState) => state.game.winner;
export const selectCanPlay = (palyeKey: keyof PlayerState | undefined) => (
  state: RootState
) => {
  if (!palyeKey) return false;
  return state.game.canPlay[palyeKey];
};

export default gameSlice.reducer;
