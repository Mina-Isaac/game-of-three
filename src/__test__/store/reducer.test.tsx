import { RootState, rootReducer } from "../../store/store";
import {
  initializeGame,
  playNextMove,
  setVs,
} from "../../store/gameSlice";
import {
  initializePlayer,
  Player,
  PlayerState,
} from "../../store/playerDetailsSlice";

const getInitialState = (initial?: RootState) =>
  rootReducer(initial as RootState, {} as any);

describe("Reducer data flow", () => {
  it("should match a snapshot", () => {
    const initialState = getInitialState();
    expect(initialState).toMatchSnapshot();
  });

  it("should initialize player with the specified player key and details", () => {
    const samplePayload: Player & { playerKey: keyof PlayerState } = {
      playerKey: "player1",
      name: "sample name",
      type: "human",
    };
    const initialState = getInitialState();
    const state = rootReducer(initialState, initializePlayer(samplePayload));
    expect(state.player[samplePayload.playerKey]?.name).toEqual(
      samplePayload.name
    );
    expect(state.player[samplePayload.playerKey]?.type).toEqual(
      samplePayload.type
    );
  });
  it("should initialize game with the passed number", () => {
    const sampleNumber = 1234567;
    const initialState = getInitialState();
    const state = rootReducer(initialState, initializeGame(sampleNumber));
    expect(state.game.baseNumber).toEqual(sampleNumber);
    expect(state.game.startingNumber).toEqual(sampleNumber);
  });
  it("should paly next move with the playNext() action", () => {
    const initialState = getInitialState();
    initialState.game.baseNumber = 8;
    initialState.game.validMove = 1;

    const state = rootReducer(initialState, playNextMove());
    expect(state.game.baseNumber).toEqual((8+1)/3);
    expect(state.game.validMove).toEqual(0);
  });
  it("should set state.vs with the passed value", () => {
    const initialState = getInitialState();
    expect(initialState.game.vs).toBe(undefined);
    const sampleValue = 'cpu';
    const state = rootReducer(initialState, setVs(sampleValue));
    expect(state.game.vs).toBe(sampleValue);
  });
});
