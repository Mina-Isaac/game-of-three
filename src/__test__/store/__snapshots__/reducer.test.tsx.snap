// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reducer data flow should match a snapshot 1`] = `
Object {
  "game": Object {
    "baseNumber": undefined,
    "canPlay": Object {
      "player1": false,
      "player2": true,
    },
    "lastPlayer": "player1",
    "startingNumber": undefined,
    "validMove": undefined,
    "vs": undefined,
    "winner": undefined,
  },
  "player": Object {
    "player1": undefined,
    "player2": undefined,
  },
}
`;
